core/cap-0034.md

```diff
@@ -118,86 +131,334 @@ older protocols' behavior, and continue to use old behavior until after
+the new last ledger close time.  `T1` therefore returns `txTOO_LATE`, and its
+operations are not applied.  However, as it reached ledger close before being
+failed, it consumes a sequence number (and is charged a fee).
+
+6. `T2` gets into an externalized transaction set (possibly, but not
+necessarily, the same one that `T1` got into) before it too expires.  It
+has the correct sequence number, because `T1` consumed a sequence number.
+`T2`'s operations are therefore applied -- even though its preconditions were
+intended to include guarantees that `T1`'s operations had ensured, so the
+actual postconditions of `T2` could violate any of the smart contract's intended
+invariants which had depended upon `T1`'s operations having succeeded before
+`T2`'s operations could be applied.
 
-## Design Rationale
+So `T` consumes a sequence number (and is charged a fee) but is never applied.
+There is a design pattern in some smart contracts which breaks, allowing
+inconsistent transactions to be committed, if it encounters this race:  a smart
+contract might submit a chain of transactions, with incrementing sequence
+numbers, each intended to be applied only if the previous ones succeeded.  If
+there were transactions that were intended to be constrained by the incrementing
+sequence numbers only to be applied if `T` succeeded, they could do so because
+`T` had consumed its sequence number, even though it was never applied. The
+dependent transactions would then perform operations which the smart contract
+had intended to be performed only if `T` had succeeded. This would
+potentially be arbitrarily bad for the smart contract.
+
+With the behavior proposed in this CAP, step #3 -- a node building a transaction
+set containing `T1` -- would only occur if `T1` were not expired with respect to
+the `closeTime` in the same proposed `StellarValue` as that transaction set.
+Otherwise, `T1` would not make it into a transaction set, and would therefore
+never consume a sequence number (or be charged a fee), and `T2` would fail
+validation with a bad sequence number, as the smart contract intended in that
+case.  If `T1` did make it into a transaction set, then, under the new behavior
+in this CAP, step #5 would change -- if the transaction set containing `T1` won
+nomination, then the `closeTime` of the new ledger would be the `closeTime` from
+the same `StellarValue` as `T1`, with respect to which `T1` is in this case not
+expired.  Hence, `T1` would be applied during ledger close; it would not return
+`txTOO_LATE`.  `T2` would therefore have the opportunity to be applied (assuming
+it was valid in the other respects in addition to its sequence number), and in
+this case that would be as expected, as `T1` had previously succeeded.
+
+Note that the duration that this race window in the current protocol remains
+open can be lengthened arbitrarily by the increase of any _one_ candidate
+`closeTime`, since the current protocol combines candidate `closeTime`s by
```

Comment:
this is incorrect: in order to be considered valid, a close time has to be between lcl.ct+1 second and now+1 minute

Reply:
Aha, I didn't notice that limit in `HerderSCPDriver::checkCloseTime()`.  I've corrected the document.

Reply:
And commits have been re-squashed.